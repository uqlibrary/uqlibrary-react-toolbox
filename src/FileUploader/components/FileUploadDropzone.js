import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import FileUploadDropzoneStaticContent from './FileUploadDropzoneStaticContent';

class FileUploadDropzone extends PureComponent {
    static propTypes = {
        onDrop: PropTypes.func.isRequired,
        maxSize: PropTypes.number.isRequired,
        locale: PropTypes.object.isRequired,
        disabled: PropTypes.bool,
        fileNameRestrictions: PropTypes.instanceOf(RegExp)
    };

    constructor(props) {
        super(props);
        this.dropzoneRef = null;
        this.errors = new Map([]);
    }

    /**
     * Get the list of folders using FileReader API
     *
     * @param accepted files and/or folders
     * @returns {Promise.<*>}
     */
    getDroppedFolders(accepted) {
        const acceptedFilesAndFolders = [...accepted];
        return Promise.all(
            acceptedFilesAndFolders.map(file => {
                return new Promise(resolve => {
                    const fileReader = new FileReader();
                    fileReader.onerror = () => resolve(file.name);
                    fileReader.onload = () => resolve();
                    const slice = file.slice(0, 10);
                    fileReader.readAsDataURL(slice);
                });
            })
        );
    }

    /**
     * Diff of two array
     *
     * @param files1
     * @param files2
     * @returns Array
     * @private
     */
    difference = (files1, files2) => {
        const set1 = new Set(files1);
        const set2 = new Set(files2);

        const difference = new Set([...set1].filter(file => !set2.has(file)));

        return [...difference];
    };

    filterFilesWithInvalidNames = (files) => {
        const filesToFilter = [...files];
        // const regex = new RegExp(this.props.fileNameRestrictions, 'gi');
        /*
         * Validate accepted files and get list of invalid files (check fileName, fileNameLength)
         */
        return filesToFilter.filter(file => {
            return new RegExp(this.props.fileNameRestrictions, 'gi').test(file.name);
        });
    };

    /**
     * Filter accepted files from dropzone
     *  - Remove folders and set error for folders
     *  - Remove files with invalid names and set error
     *  - Hand over valid files only to file uploader
     *
     * @param accepted
     * @param folders
     */
    filterOnDrop = (accepted, folders) => {
        const filesWithoutFolders = folders.length > 0 ? accepted.filter(file => folders.indexOf(file.name) === -1) : accepted;

        const filesWithValidNames = this.filterFilesWithInvalidNames([...filesWithoutFolders]);

        const invalidFiles = this.difference(filesWithoutFolders, filesWithValidNames);

        this.errors.set('folder', folders);
        this.errors.set('fileName', invalidFiles.map(file => file.name));

        this.props.onDrop([...filesWithValidNames], this.errors);

        this.errors = new Map([]);
    };

    /**
     * Handle accepted and rejected files on dropped in Dropzone
     *
     * @param accepted
     * @param rejected
     * @param event
     * @private
     */
    _onDrop = (accepted, rejected, event) => {
        this.errors.set('maxFileSize', rejected.map(file => file.name));

        /*
         * From droppedEvent dataTransfer items, determine which items are folders
         *
         * Safari and IE doesn't support event.dataTransfer.items
         * https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/items
         *
         * Using FileReader API async to read slice of file will throw an error if it's a folder
         */
        if (!!event && !!event.dataTransfer && !!event.dataTransfer.items) {
            const folders =  Array.prototype.filter.call(event.dataTransfer.items, (item) => (item.webkitGetAsEntry().isDirectory))
                .map((item) => item.webkitGetAsEntry().name);

            this.filterOnDrop([...accepted], [...folders]);
        } else {
            this.getDroppedFolders([...accepted]).then(result => {
                const folders = result.filter(folder => !!folder);

                this.filterOnDrop([...accepted], [...folders]);
            });
        }
    };

    /**
     * Open dropzone on key pressed
     */
    _onKeyPress = () => {
        this.dropzoneRef.open();
    };

    render() {
        const {maxSize, disabled, locale} = this.props;
        return (
            <div>
                <div className="columns">
                    <div className="column"  tabIndex="0" onKeyPress={this._onKeyPress}>
                        <Dropzone
                            ref={(node) => {this.dropzoneRef = node;}}
                            maxSize={maxSize}
                            onDrop={this._onDrop}
                            style={{padding: '0px'}}
                            disabled={disabled}
                            disableClick={disabled}
                            disablePreview>
                            <FileUploadDropzoneStaticContent txt={locale}/>
                        </Dropzone>
                    </div>
                </div>
            </div>
        );
    }
}

export default FileUploadDropzone;

