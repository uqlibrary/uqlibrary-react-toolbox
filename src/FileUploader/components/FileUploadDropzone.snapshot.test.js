import FileUploadDropzone from './FileUploadDropzone';

const locale = {
    validation: {
        ['folder']: 'Invalid files ([filenames])',
        ['fileName']: 'File(s) ([filenames]) have invalid file name',
        ['maxFileSize']: 'File(s) ([filenames]) exceed maximum allowed upload file size',
        ['maxFiles']: 'Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded',
    },
    successTitle: 'Success',
    successMessage: 'Successfully added [numberOfFiles] file(s) to upload queue.',
    errorTitle: 'Upload Errors',
    fileUploadRestrictionHeading: 'Test header',
    fileUploadRestrictions: 'Test text',
    fileUploadInstruction: "Test instructions"
};

function setup(testProps, isShallow = true) {
    const props = {
        onDropped: jest.fn(),
        maxSize: 1000,
        maxFiles: 5,
        uploadedFiles: [],
        clearErrors: false,
        locale: locale,
        ...testProps,
    };
    return getElement(FileUploadDropzone, props, isShallow);
}

describe('Component FileUploadDropzone', () => {
    beforeEach(() => {
        const FILE_READER_TO_USE = new FileReader();
        window.FileReader = jest.fn(() => FILE_READER_TO_USE);
        window.FileReader.onerror = () => resolve();
        window.FileReader.onload = () => resolve();
        window.FileReader.readAsDataURL = () => (window.FileReader.onload);
    });

    it('should render correctly without any setup', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render files discarding folders', () => {
        const onDroppedCallback = jest.fn();
        const props = {
            onDropped: onDroppedCallback
        };
        const wrapper = setup({...props});

        expect(toJson(wrapper)).toMatchSnapshot();

        const accepted = [
            {
                type: 'text/text',
                name: 'a.txt',
                size: 500
            },
            {
                type: 'text/text',
                name: 'a.text.txt',
                size: 100
            },
            {
                type: 'text/text',
                name: 'ab.txt',
                size: 100
            },
            {
                type: '',
                name: 'test',
                size: 100
            },
            {
                type: 'text/text',
                name: 'web_a.txt',
                size: 100
            },
            {
                type: 'text/text',
                name: 'WEB_b.txt',
                size: 100
            },
            {
                type: 'text/text',
                name: 'Web_c.txt',
                size: 100
            }
        ];

        const event = {
            dataTransfer: {
                items: [
                    {
                        webkitGetAsEntry: () => ({
                            name: 'test',
                            isDirectory: true
                        })
                    }
                ]
            }
        };

        wrapper.instance()._onDrop(accepted, [], event);
        wrapper.update();

        expect(toJson(wrapper)).toMatchSnapshot();
        expect(onDroppedCallback).toHaveBeenCalled();
    });

    it('should set max file size error message for rejected files', () => {
        const wrapper = setup({});

        const accepted = [
            {
                name: 'a.txt',
                size: 500
            }
        ];

        const rejected = [
            {
                name: 'b.txt',
                size: 10000
            }
        ];

        const event = {
            dataTransfer: {
                items: []
            }
        };

        wrapper.instance()._onDrop(accepted, rejected, event);
        wrapper.update();
        expect(wrapper.state().errorMessage).toEqual('File(s) (b.txt) exceed maximum allowed upload file size');
    });

    it('should set max files error message', () => {
        const wrapper = setup({maxFiles: 3});
        wrapper.instance().componentWillReceiveProps({
            uploadedFiles: [
                {
                    name: 'a.txt',
                    size: 100
                },
                {
                    name: 'b.txt',
                    size: 100
                }
            ]
        });

        const accepted = [
            {
                name: 'c.txt',
                size: 500
            },
            {
                name: 'd.txt',
                size: 10000
            }
        ];

        const event = {
            dataTransfer: {
                items: []
            }
        };

        wrapper.instance()._onDrop(accepted, [], event);
        wrapper.update();
        expect(wrapper.state().errorMessage).toEqual('Maximum number of files (3) has been exceeded. File(s) (d.txt) will not be uploaded');
    });

    it('should open files selection dialog', () => {
        const onDroppedCallback = jest.fn();
        const props = {
            onDropped: onDroppedCallback,
            maxSize: 1000,
            maxFiles: 5,
            uploadedFiles: [],
            clearErrors: false,
            locale: locale
        };
        const wrapper = setup({...props}, false);

        expect(toJson(wrapper)).toMatchSnapshot();

        const testFn = jest.fn();

        wrapper.find('FileUploadDropzone').instance().dropzoneRef.open = testFn;
        wrapper.find('FileUploadDropzone').instance()._onKeyPress();

        wrapper.update();
        expect(testFn).toHaveBeenCalled();
    });

    it('should set success alert for number of uploaded files', () => {
        const props = {
            maxFiles: 5,
            uploadedFiles: [],
            locale: locale
        };

        const files = new Set([
            {
                name: 'c.txt',
                size: 100
            },
            {
                name: 'd.txt',
                size: 100
            },
            {
                name: 'd.txt',
                size: 100
            }
        ]);

        const wrapper = setup({...props});
        wrapper.state({successMessage: 'Successfully added 2 file(s) to upload queue.'});
        wrapper.instance().componentWillReceiveProps({
            uploadedFiles: [
                {
                    name: 'a.txt',
                    size: 100
                },
                {
                    name: 'b.txt',
                    size: 100
                }
            ],
            clearErrors: true
        });

        wrapper.instance().setSuccessMessage(files);
        wrapper.update();
        expect(wrapper.state().successMessage).toEqual('Successfully added 3 file(s) to upload queue.');
    });

    it('should clear success alert if max number of files already added', () => {
        const props = {
            maxFiles: 3,
            uploadedFiles: [],
            locale: locale
        };

        const files = new Set([
            {
                name: 'd.txt',
                size: 100
            }
        ]);

        const wrapper = setup({...props});

        wrapper.state({successMessage: 'Successfully added 3 file(s) to upload queue.'});
        wrapper.instance().componentWillReceiveProps({
            uploadedFiles: [
                {
                    name: 'a.txt',
                    size: 100
                },
                {
                    name: 'b.txt',
                    size: 100
                },
                {
                    name: 'c.txt',
                    size: 100
                }
            ]
        });
        wrapper.instance().setSuccessMessage(files);
        wrapper.update();
        expect(wrapper.state().successMessage).toEqual('');
    });

    it('should set success message for max number of files if files count is greater than max files', () => {

        const props = {
            maxFiles: 3,
            uploadedFiles: [],
            locale: locale
        };

        const files = new Set([
            {
                name: 'a.txt',
                size: 100
            },
            {
                name: 'b.txt',
                size: 100
            },
            {
                name: 'c.txt',
                size: 100
            },
            {
                name: 'd.txt',
                size: 100
            }
        ]);

        const wrapper = setup({...props});
        wrapper.instance().setSuccessMessage(files);
        expect(wrapper.state().successMessage).toEqual('Successfully added 3 file(s) to upload queue.');
    });
});
