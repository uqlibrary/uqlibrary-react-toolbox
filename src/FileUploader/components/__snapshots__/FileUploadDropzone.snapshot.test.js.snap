// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component FileUploadDropzone should open files selection dialog 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    }
  }
>
  <FileUploadDropzone
    clearErrors={false}
    fileNameRestrictions={/\\^\\(\\?=\\^\\\\S\\*\\$\\)\\(\\?=\\^\\[\\^\\\\\\.\\]\\+\\\\\\.\\[\\^\\\\\\.\\]\\+\\$\\)\\(\\?=\\.\\{1,45\\}\\$\\)\\(\\?!\\(web_\\|preview_\\|thumbnail_\\|stream_\\|fezacml_\\|presmd_\\)\\)\\[a-z\\]\\[a-z\\\\d\\\\-_\\\\\\.\\]\\+/}
    locale={
      Object {
        "errorTitle": "Upload Errors",
        "fileUploadInstruction": "Test instructions",
        "fileUploadRestrictionHeading": "Test header",
        "fileUploadRestrictions": "Test text",
        "successMessage": "Successfully added [numberOfFiles] file(s) to upload queue.",
        "successTitle": "Success",
        "validation": Object {
          "fileName": "File(s) ([filenames]) have invalid file name",
          "folder": "Invalid files ([filenames])",
          "maxFileSize": "File(s) ([filenames]) exceed maximum allowed upload file size",
          "maxFiles": "Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded",
        },
      }
    }
    maxFiles={5}
    maxSize={1000}
    onDropped={[MockFunction]}
    uploadedFiles={Array []}
  >
    <div>
      <div
        className="columns"
      >
        <div
          className="column"
          onKeyPress={[Function]}
          tabIndex="0"
        >
          <t
            disableClick={false}
            disablePreview={true}
            disabled={false}
            maxSize={1000}
            minSize={0}
            multiple={true}
            onDrop={[Function]}
            preventDropOnDocument={true}
            style={
              Object {
                "padding": "0px",
              }
            }
          >
            <div
              aria-disabled={false}
              className=""
              onClick={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              style={
                Object {
                  "padding": "0px",
                }
              }
            >
              <FileUploadDropzoneStaticContent
                txt={
                  Object {
                    "errorTitle": "Upload Errors",
                    "fileUploadInstruction": "Test instructions",
                    "fileUploadRestrictionHeading": "Test header",
                    "fileUploadRestrictions": "Test text",
                    "successMessage": "Successfully added [numberOfFiles] file(s) to upload queue.",
                    "successTitle": "Success",
                    "validation": Object {
                      "fileName": "File(s) ([filenames]) have invalid file name",
                      "folder": "Invalid files ([filenames])",
                      "maxFileSize": "File(s) ([filenames]) exceed maximum allowed upload file size",
                      "maxFiles": "Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded",
                    },
                  }
                }
              >
                <div
                  className="columns file-instructions"
                >
                  <div
                    className="column"
                  >
                    Test header
                    Test text
                  </div>
                  <div
                    className="column upload-instructions"
                  >
                    <FontIcon
                      className="material-icons"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="material-icons"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        style={
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "display": "inline-block",
                            "fontSize": 24,
                            "muiPrepared": true,
                            "position": "relative",
                            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                            "userSelect": "none",
                          }
                        }
                      >
                        cloud_upload
                      </span>
                    </FontIcon>
                    Test instructions
                  </div>
                </div>
              </FileUploadDropzoneStaticContent>
              <input
                autoComplete="off"
                disabled={false}
                multiple={true}
                onChange={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="file"
              />
            </div>
          </t>
        </div>
      </div>
    </div>
  </FileUploadDropzone>
</Provider>
`;

exports[`Component FileUploadDropzone should render correctly without any setup 1`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="column"
      onKeyPress={[Function]}
      tabIndex="0"
    >
      <t
        disableClick={false}
        disablePreview={true}
        disabled={false}
        maxSize={1000}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "padding": "0px",
          }
        }
      >
        <FileUploadDropzoneStaticContent
          txt={
            Object {
              "errorTitle": "Upload Errors",
              "fileUploadInstruction": "Test instructions",
              "fileUploadRestrictionHeading": "Test header",
              "fileUploadRestrictions": "Test text",
              "successMessage": "Successfully added [numberOfFiles] file(s) to upload queue.",
              "successTitle": "Success",
              "validation": Object {
                "fileName": "File(s) ([filenames]) have invalid file name",
                "folder": "Invalid files ([filenames])",
                "maxFileSize": "File(s) ([filenames]) exceed maximum allowed upload file size",
                "maxFiles": "Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded",
              },
            }
          }
        />
      </t>
    </div>
  </div>
</div>
`;

exports[`Component FileUploadDropzone should render files discarding folders 1`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="column"
      onKeyPress={[Function]}
      tabIndex="0"
    >
      <t
        disableClick={false}
        disablePreview={true}
        disabled={false}
        maxSize={1000}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "padding": "0px",
          }
        }
      >
        <FileUploadDropzoneStaticContent
          txt={
            Object {
              "errorTitle": "Upload Errors",
              "fileUploadInstruction": "Test instructions",
              "fileUploadRestrictionHeading": "Test header",
              "fileUploadRestrictions": "Test text",
              "successMessage": "Successfully added [numberOfFiles] file(s) to upload queue.",
              "successTitle": "Success",
              "validation": Object {
                "fileName": "File(s) ([filenames]) have invalid file name",
                "folder": "Invalid files ([filenames])",
                "maxFileSize": "File(s) ([filenames]) exceed maximum allowed upload file size",
                "maxFiles": "Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded",
              },
            }
          }
        />
      </t>
    </div>
  </div>
</div>
`;

exports[`Component FileUploadDropzone should render files discarding folders 2`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="column"
      onKeyPress={[Function]}
      tabIndex="0"
    >
      <t
        disableClick={false}
        disablePreview={true}
        disabled={false}
        maxSize={1000}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "padding": "0px",
          }
        }
      >
        <FileUploadDropzoneStaticContent
          txt={
            Object {
              "errorTitle": "Upload Errors",
              "fileUploadInstruction": "Test instructions",
              "fileUploadRestrictionHeading": "Test header",
              "fileUploadRestrictions": "Test text",
              "successMessage": "Successfully added [numberOfFiles] file(s) to upload queue.",
              "successTitle": "Success",
              "validation": Object {
                "fileName": "File(s) ([filenames]) have invalid file name",
                "folder": "Invalid files ([filenames])",
                "maxFileSize": "File(s) ([filenames]) exceed maximum allowed upload file size",
                "maxFiles": "Maximum number of files ([maxNumberOfFiles]) has been exceeded. File(s) ([filenames]) will not be uploaded",
              },
            }
          }
        />
      </t>
    </div>
  </div>
  <Alert
    allowDismiss={false}
    message="Successfully added 2 file(s) to upload queue."
    title="Success"
    type="done"
  />
  <Alert
    allowDismiss={false}
    message="Invalid files (test); File(s) (a.text.txt, web_a.txt, WEB_b.txt, Web_c.txt) have invalid file name"
    title="Upload Errors"
    type="error"
  />
</div>
`;
